src/pykeen/sampling/basic_negative_sampler.py

```diff
@@ -31,31 +34,84 @@ class BasicNegativeSampler(NegativeSampler):
         num_negs_per_pos=dict(type=int, low=1, high=100, q=10),
     )
 
-    def sample(self, positive_batch: torch.LongTensor) -> torch.LongTensor:
+    def __init__(
+        self,
+        triples_factory: TriplesFactory,
+        num_negs_per_pos: Optional[int] = None,
+        filtered: bool = False,
+        corruption_scheme: Optional[Collection[str]] = None,
+    ) -> None:
+        """Initialize the negative sampler with the given entities.
+
+        :param triples_factory: The factory holding the triples to sample from
+        :param num_negs_per_pos: Number of negative samples to make per positive triple. Defaults to 1.
+        :param filtered: Whether proposed corrupted triples that are in the training data should be filtered.
+            Defaults to False.
+        :param corruption_scheme: What sides ('h', 'r', 't') should be corrupted. Defaults to head and tail ('h', 't').
```

Comment:
Did the reviewer have a specific use case when you might want to do something other than default?

Reply:
No, but there could be reasonable situations, e.g. when working with images where you know the objects/entities, but are interested in training to reason the most likely relation.

Reply:
I cannot find a reference right now, but iirc visual relation detection usually first detects all objects in the image (`h` / `t`) and then predicts the relations between them (`r`). One argument is that the distribution of visual representations of relations is more of a long tail distribution, and might include representations of the objects: e.g. in `(man, rides, bike)`' the visual representations of `rides` likely (partially) shows both, the man and the bike.

Far better explanations can be given by @sharifza 🙂